scripts/build_nim.sh

```diff
@@ -18,7 +18,8 @@ set -e
 
 # After this Nim commit, use csources v2
 : ${CSOURCES_V2_START_COMMIT:=f7c203fb6c89b5cef83c4f326aeb23ef8c4a2c40}
-: ${NIMBLE_COMMIT:=d13f3b8ce288b4dc8c34c219a4e050aaeaf43fc9} # 0.13.1
+# Set Nimble commit in environment to have non default one.
+# e.g. NIMBLE_COMMIT=3575fd54a890d910ace56678aa74b4237d604175 # 0.14.2
```

Comment:
Last time I checked, was advised that the Nimble version is tied to the Nim version (it hardcodes a specific commit to be used). So, not sure how functional overriding is. If there is a use case, it looks quite clean.

Reply:
I am aware there is a relationship between Nim and Nimble versions. 

I don't think however it is something which could create much of problems. At least when I tested this with nim-libp2p https://github.com/vacp2p/nim-libp2p/pull/1004 I haven't had any problem. Right after that I opened this PR to contribute back. 

PR 1004 was rejected based on this generic advice without any further explanation. 

Nimble v0.14.2 is out more than 1 year. Entire Nim ecosystem might be behind similar competitors (Go, Rust) about several years. 

Nimbus is the end2end application for Ethereum ecosystem. Single producer - IFT delivers to many consumers - validators. This is completely different with Waku and Codex, where IFT is a major producer of middleware expecting adoption by integrators. Nim ecosystem quality, dev experience, quality of libs directly affects adoption of Waku and Codex. 

Nimbus has about 6% market share among Ethereum validators. While I do admire the achievement, I am far from thrilled about this number.

There is a substantial chance Waku and Codex potential will be capped by Nim ecosystem mediocrity. Especially when the scarcity mindset is taken for Nim development by a major Nim ecosystem contributor. Mediocre (at best) outcome is almost guaranteed. 

Having said all the negative, there is an another approach main Nim contributor might take which can substantially improve odds for Waku and Codex (and potentially for Nimbus too) to make it big:

- accountability: there must be someone working full time on compiler and tooling. Someone who is praised, blamed, hired or fired because of quality of their work. Tinkeners or hobbyist part time approach won't work. Sorry for those part timers out there. You can't be the steering force. 
- separation of responsibilities: Waku team builds Waku and complains or praises to Nim eco team. No compiler builds and similar repetitive slow downs.
- clearly defined feedback loop for Nim eco dev

Bad dev experience scales exponentially into silence of those who tried and moved hands out. Good dev experience scales also exponentially, but into growth and positive impact!

Have a chat or meeting with me, feel free to dispute or challenge in productive manner.

Reply:
Generically, could Waku and Codex use Nim 2.0 at this point (using `refc` initially probably, since some libraries such as `nim-metrics` don't work well in ORC yet)? 2.0.4 might be better quality than 1.6.x and 2.0.6, hopefully soon, should be fairly solid.

It reports
```
$ bin/nimble --version
nimble v0.14.2 compiled at 2024-06-16 18:41:44
git hash: f8bd7b5fa6ea7a583b411b5959b06e6b5eb23667
```

Reply:
fwiw, it's broadly an aim to start disconnecting nim from nimble and have the latter install the former - disconnecting their versions here is a good first step and indeed, it makes sense to move faster on nimble now that we're investing in its development.

something that needs doing for moving the needle here is to test this bump with a number of major nimble projects such as libp2p, chronos and maybe a few others from `nimbus-eth2/vendor` - changing the nimble here should also result in _their_ ci:s starting to use nimble this way - there's a bit of work to do here, though it's fairly straightforward - it includes bumping some of _those_ nimbles to newer versions - what we're looking for is to ideally use a similar version across projects.

In an ideal world, we'd actually have a github action that can download nimble - this nimble would be used to install nim (instead of having the action install nim and its included, out-of-date version.


cc @jmgomez

Reply:
> Generically, could Waku and Codex use Nim 2.0 at this point (using `refc` initially probably, since some libraries such as `nim-metrics` don't work well in ORC yet)? 2.0.4 might be better quality than 1.6.x and 2.0.6, hopefully soon, should be fairly solid.
> 
> It reports
> 
> ```
> $ bin/nimble --version
> nimble v0.14.2 compiled at 2024-06-16 18:41:44
> git hash: f8bd7b5fa6ea7a583b411b5959b06e6b5eb23667
> ```

Indeed a valid option to move to Nim 2.0. Thanks for mentioning it.

Reply:
> fwiw, it's broadly an aim to start disconnecting nim from nimble and have the latter install the former - disconnecting their versions here is a good first step and indeed, it makes sense to move faster on nimble now that we're investing in its development.
> 
> something that needs doing for moving the needle here is to test this bump with a number of major nimble projects such as libp2p, chronos and maybe a few others from `nimbus-eth2/vendor` - changing the nimble here should also result in _their_ ci:s starting to use nimble this way - there's a bit of work to do here, though it's fairly straightforward - it includes bumping some of _those_ nimbles to newer versions - what we're looking for is to ideally use a similar version across projects.
> 
> In an ideal world, we'd actually have a github action that can download nimble - this nimble would be used to install nim (instead of having the action install nim and its included, out-of-date version.
> 
> cc @jmgomez

This is a great idea, let me know if I can help. What criteria for the "test after bump" should be checked ? If there is no specific list, I was thinking even my testing was successful with nim-libp2p, we might still need some time to give problems pop up. Something like two weeks with new Nimble and at least one team member per project to try ?

Reply:
https://nim-lang.org/blog/2024/06/17/version-206-released.html

Reply:
@romanzac  well, we can experiment on https://github.com/status-im/nim-stew/ for example - make a second CI job that runs its tests on nim 1.6, 2.0 and some nightly version except now, it should download a binary nimble and install Nim using that.

libp2p has too complex requirements but also is not well set up for this kind of experiment - many of our libraries, we want to test them with a diverse set of dependencies - including nim - libraries are more useful if they are flexible in their version constraints, so this is a good litmus test - stew is also easy enough in that it doesn't have many deps.

We should start with a nightly nimble that has the new sat solver, else we won't get anywhere.

Reply:
> @romanzac well, we can experiment on https://github.com/status-im/nim-stew/ for example - make a second CI job that runs its tests on nim 1.6, 2.0 and some nightly version except now, it should download a binary nimble and install Nim using that.
> 
> libp2p has too complex requirements but also is not well set up for this kind of experiment - many of our libraries, we want to test them with a diverse set of dependencies - including nim - libraries are more useful if they are flexible in their version constraints, so this is a good litmus test - stew is also easy enough in that it doesn't have many deps.
> 
> We should start with a nightly nimble that has the new sat solver, else we won't get anywhere.

May I have permissions to the stew repo ? I was trying to commit and play the process you have described for Linux first.

Reply:
Some issues emerged after the run https://github.com/status-im/nim-stew/actions/runs/9572076685/job/26390602852 I will go through them in coming days.

Reply:
`nimbus-eth1` has been switched to Nim 2.0.6 (and thus by default, its version of Nimble):
- https://github.com/status-im/nimbus-eth1/pull/2384

It doesn't obviate this PR's purpose overall, but it, and subsequent migrations, will alleviate some of the immediate issue.

Reply:
> `nimbus-eth1` has been switched to Nim 2.0.6 (and thus by default, its version of Nimble):
> 
> * [use Nim 2.0.6 nimbus-eth1#2384](https://github.com/status-im/nimbus-eth1/pull/2384)
> 
> It doesn't obviate this PR's purpose overall, but it, and subsequent migrations, will alleviate some of the immediate issue.

Good sharing and good point. We should probably coordinate in Discord or more convenient format. This PR is more like a forum now. 

1) I am trying to make nominal case to run in CI with latest Nimble and Nim for Nim Stew. Got one failed test only, not bad at all. Will update you tomorrow. 
2) I already know while playing on local linux box, latest Nimble won't let me install Nim 1.6.

I think moving all IFT's projects to Nim 2.0 is inevitable. Tooling will get upgraded automatically. Question is when and how to process problems. 

Second thought: When we picked Nim as to go lang, we implicitly accepted "other" Nim contributors to be both with us and in some cases not in line with us. That means, we can only exercise control by pulling the rest of Nim community with us. Pushing them won't work longterm. Pulling the community requires horses. :)  I know some of our best devs could produce 5 substantial commits per week. Latest Nim release mentions 200/2 months. That is 25 commits per week. Talking in IFT staff numbers we may say 5 team members work? Not sure how accurate my guess is, we might both overwhelmingly alter Nim lang direction to our liking and ignite broader community interest. We could invest gradually and see how it goes. Once there is enough momentum we may invest less or more. Nim could be go to lang for networking, security and privacy solutions. Sky is the limit ?! :) 

Sorry to be so expressive :) Let coordinate ourselves first...

Reply:
Sorry I had no time on Friday. Nominal case for all tests run finished with 0 failed tests:

[Summary] 150 tests run (0.77s): 148 OK, 0 FAILED, 2 SKIPPED

https://github.com/status-im/nim-stew/actions/runs/9651769206/job/26620400645

I'll move forward with other combinations tomorrow. Potentially also opening issue for 1.6 installation.

Reply:
`nimbus-eth2` migrated to Nim 2.0.6: https://github.com/status-im/nimbus-eth2/pull/6386

Reply:
Waku migrated to Nim 2.0.8 (and `nimbus-eth1` and `nimbus-eth2` have as well, but at this point it's just a minor version change for them): https://github.com/waku-org/nwaku/pull/2885

Reply:
Good news Waku have migrated! We need to sustain this speed, quality and focus on Nim eco related development to create momentum. This would eventually draw attention of very capable non CC contributors. And we will literally multiply fruits of our effort by many folds.